File: frontend/app/utils/password.ts
export function isValidPassword(password: string) {
    let errorMessage = '';

    // Check length
    if (password.length < 8 || password.length > 64) {
        errorMessage = 'Password must be between 8-64 characters long.';
        return { isValid: false, errorMessage };
    }

    // Check complexity
    const hasUppercase = /[A-Z]/.test(password);
    const hasLowercase = /[a-z]/.test(password);
    const hasNumber = /[0-9]/.test(password);
    const hasSpecialChar = /[!@#$%^&*(),.?":{}|<>]/.test(password);

    if (!hasUppercase) {
        errorMessage = 'Password must contain at least 1 uppercase letter (A-Z).';
        return { isValid: false, errorMessage };
    }

    if (!hasLowercase) {
        errorMessage = 'Password must contain at least 1 lowercase letter (a-z).';
        return { isValid: false, errorMessage };
    }

    if (!hasNumber) {
        errorMessage = 'Password must contain at least 1 numeral (0-9).';
        return { isValid: false, errorMessage };
    }

    if (!hasSpecialChar) {
        errorMessage = 'Password must contain at least 1 special character.';
        return { isValid: false, errorMessage };
    }

    // Check for spaces
    if (/\s/.test(password)) {
        errorMessage = 'Password must not contain any spaces.';
        return { isValid: false, errorMessage };
    }

    return { isValid: true, errorMessage };
}

File: api/internal/pkg/utils/password.go
package utils

import (
	"errors"
	"regexp"
	"strings"
)

func IsValidPassword(password string) (bool, error) {
	// Check length
	if len(password) < 8 || len(password) > 64 {
		return false, errors.New("password must be between 8-64 characters long.")
	}

	// Check complexity
	hasUppercase, _ := regexp.MatchString(`[A-Z]`, password)
	hasLowercase, _ := regexp.MatchString(`[a-z]`, password)
	hasNumber, _ := regexp.MatchString(`[0-9]`, password)
	hasSpecialChar, _ := regexp.MatchString(`[!@#$%^&*(),.?":{}|<>]`, password)

	if !hasUppercase {
		return false, errors.New("password must contain at least 1 uppercase letter (A-Z).")
	}

	if !hasLowercase {
		return false, errors.New("password must contain at least 1 lowercase letter (a-z).")
	}

	if !hasNumber {
		return false, errors.New("password must contain at least 1 numeral (0-9).")
	}

	if !hasSpecialChar {
		return false, errors.New("password must contain at least 1 special character.")
	}

	// Check for spaces
	if strings.Contains(password, " ") {
		return false, errors.New("password must not contain any spaces.")
	}

	return true, nil
}
